src/components/LocaleContextProvider.tsx

```diff
@@ -0,0 +1,147 @@
+    /** The user's preferred locale e.g. 'en', 'es-ES' */
+    preferredLocale: string;
+};
+
+type LocaleContextProviderProps = LocaleContextProviderOnyxProps & {
+    /** The current user's personalDetails */
+    currentUserPersonalDetails: CurrentUserPersonalDetails;
+
+    /** Actual content wrapped by this component */
+    children: React.ReactNode;
+};
+
+type Translate = (phrase: string, variables: Record<string, string>) => string;
+
+type NumberFormat = (number: number, options: Intl.NumberFormatOptions) => string;
+
+type DatetimeToRelative = (datetime: string) => string;
+
+type DatetimeToCalendarTime = (datetime: string, includeTimezone: boolean, isLowercase: boolean) => string;
+
+type UpdateLocale = () => void;
+
+type FormatPhoneNumber = (phoneNumber: string) => string;
+
+type ToLocaleDigit = (digit: string) => string;
+
+type FromLocaleDigit = (digit: string) => string;
+
+type LocaleContextProps = {
+    /** Returns translated string for given locale and phrase */
+    translate: Translate;
+
+    /** Formats number formatted according to locale and options */
+    numberFormat: NumberFormat;
+
+    /** Converts a datetime into a localized string representation that's relative to current moment in time */
+    datetimeToRelative: DatetimeToRelative;
+
+    /** Formats a datetime to local date and time string */
+    datetimeToCalendarTime: DatetimeToCalendarTime;
+
+    /** Updates date-fns internal locale */
+    updateLocale: UpdateLocale;
+
+    /** Returns a locally converted phone number for numbers from the same region
+     * and an internationally converted phone number with the country code for numbers from other regions */
+    formatPhoneNumber: FormatPhoneNumber;
+
+    /** Gets the locale digit corresponding to a standard digit */
+    toLocaleDigit: ToLocaleDigit;
+
+    /** Gets the standard digit corresponding to a locale digit */
+    fromLocaleDigit: FromLocaleDigit;
+
+    /** The user's preferred locale e.g. 'en', 'es-ES' */
+    preferredLocale: string;
+};
+
+const LocaleContext = createContext<LocaleContextProps>({
+    translate: () => '',
+    numberFormat: () => '',
+    datetimeToRelative: () => '',
+    datetimeToCalendarTime: () => '',
+    updateLocale: () => '',
+    formatPhoneNumber: () => '',
+    toLocaleDigit: () => '',
+    fromLocaleDigit: () => '',
+    preferredLocale: CONST.LOCALES.DEFAULT,
+});
+
+function LocaleContextProvider({preferredLocale = CONST.LOCALES.DEFAULT, currentUserPersonalDetails = {}, children}: LocaleContextProviderProps) {
+    const selectedTimezone = useMemo(() => currentUserPersonalDetails?.timezone?.selected, [currentUserPersonalDetails]);
+
+    const translate = useMemo<Translate>(() => (phrase, variables) => Localize.translate(preferredLocale, phrase, variables), [preferredLocale]);
+
+    const numberFormat = useMemo<NumberFormat>(() => (number, options) => NumberFormatUtils.format(preferredLocale, number, options), [preferredLocale]);
```

Comment:
I noticed that some of these util functions (lines 95-116) accept strings, this is something we haven't noticed before, so I'd recommend going through all of them and making sure that they are strict enough - in cases where we use timezones, make use of that new type, and where we use LOCALES make use of ValueOf<typeof CONST.LOCALES>